File: paleo_workbench/workflow/dag/store.py
"""WorkflowRun persistence (H3): atomic JSON checkpoints in project-managed
storage, plus the deterministic cache lookup over prior checkpoints.

The store never invents output truth: a cache hit requires (a) an identical
cache identity (action id+version, normalized parameters, input version
ids), (b) a terminal-successful node, and (c) every recorded output version
resolvable in the catalog with a passing integrity check (the catalog stays
the only artifact authority). Anything less re-executes.
"""
from __future__ import annotations

import json
import logging
import os
import tempfile
import time
from pathlib import Path
from typing import Any

from paleo_workbench.workflow.dag.model import NodeRun, RunState, WorkflowRun

logger = logging.getLogger(__name__)

_STORE_VERSION = 1


class WorkflowRunStore:
    """File-backed run store. One JSON file per run, written atomically.

    V8 M7 cache index: ``find_reusable_node`` used to JSON-parse EVERY run
    file per cacheable-node lookup (O(runs x nodes) with full
    deserialization). The store now maintains an in-memory
    ``cache_identity -> [(run_id, node_id)]`` index, built once from disk on
    first lookup and maintained incrementally on ``save`` — lookups
    deserialize only the candidate runs (newest first). The index is a
    lookup aid, never a second authority: a candidate is still re-loaded
    from its file and re-validated (state, from_cache, output resolvability)
    before reuse.
    """

    def __init__(self, root: str | Path) -> None:
        self.root = Path(root)
        self.root.mkdir(parents=True, exist_ok=True)
        self._cache_index: dict[str, list[tuple[str, str]]] | None = None

    # ------------------------------------------------------------ index --
    def _cache_index_entry(self, run: WorkflowRun) -> None:
        """Record one run's reusable nodes in the index (if built)."""
        if self._cache_index is None:
            return
        if run.state == RunState.RUNNING:
            return  # not reusable evidence until terminal
        for node_run in run.node_runs.values():
            if (
                node_run.state.value == "succeeded"
                and node_run.cache_identity
                and not node_run.from_cache
                and node_run.output_version_ids
            ):
                self._cache_index.setdefault(node_run.cache_identity, []).append(
                    (run.run_id, node_run.node_id)
                )

    def rebuild_cache_index(self) -> int:
        """Scan every stored run once and build the cache index.

        Returns the number of indexed node entries. Runs load through the
        normal path — a corrupted run is skipped (logged), never fatal.
        """
        self._cache_index = {}
        for run in self.list_runs():
            self._cache_index_entry(run)
        return sum(len(v) for v in self._cache_index.values())

    def candidates_for_identity(self, cache_identity: str) -> list[tuple[str, str]]:
        """(run_id, node_id) candidates for a cache identity, NEWEST first."""
        if self._cache_index is None:
            self.rebuild_cache_index()
        assert self._cache_index is not None
        entries = list(self._cache_index.get(cache_identity, []))
        entries.reverse()  # later saves appended last -> newest first
        return entries

    # ------------------------------------------------------------ CRUD --
    def _path(self, run_id: str) -> Path:
        return self.root / f"run-{run_id}.json"

    def save(self, run: WorkflowRun) -> Path:
        run.updated_at = time.time()
        path = self._path(run.run_id)
        payload = {"store_version": _STORE_VERSION, **run.to_dict()}
        fd, tmp = tempfile.mkstemp(dir=str(self.root), prefix=".tmp-run-", suffix=".json")
        try:
            with os.fdopen(fd, "w", encoding="utf-8") as fh:
                json.dump(payload, fh, ensure_ascii=False, indent=1)
            os.replace(tmp, path)
        except Exception:
            try:
                os.unlink(tmp)
            except OSError:
                pass
            raise
        self._cache_index_entry(run)
        return path

    def load(self, run_id: str) -> WorkflowRun:
        path = self._path(run_id)
        if not path.exists():
            raise KeyError(f"no workflow run {run_id!r} in {self.root}")
        try:
            data = json.loads(path.read_text(encoding="utf-8"))
        except json.JSONDecodeError as exc:
            # A corrupted checkpoint is never executed from: it surfaces as
            # a load failure the caller must handle (re-run from scratch).
            raise ValueError(f"workflow run {run_id!r} checkpoint is corrupted: {exc}") from exc
        return WorkflowRun.from_dict(data)

    def list_run_ids(self) -> list[str]:
        return sorted(
            p.stem.removeprefix("run-") for p in self.root.glob("run-*.json")
        )

    def list_runs(self) -> list[WorkflowRun]:
        runs = []
        for run_id in self.list_run_ids():
            try:
                runs.append(self.load(run_id))
            except Exception:  # a corrupted run must not poison cache lookups
                logger.warning("skipping unreadable workflow run %s", run_id)
        return runs


def default_store_root(context: Any) -> Path:
    """Project-managed storage root: ``<project>.artifacts/workflows/``.

    The artifacts tree is derived by the single authority
    (:func:`paleo_workbench.project.paths.artifact_dir_for`) — the same
    helper every other artifact writer uses, so workflow runs never land
    in a second tree. Falls back to a session directory when the action
    context has no project (headless unit runs).
    """
    project_path = getattr(context, "project_path", None)
    if project_path:
        from paleo_workbench.project.paths import artifact_dir_for

        root = artifact_dir_for(Path(project_path)) / "workflows"
    else:
        root = Path(tempfile.gettempdir()) / "paleo-workflow-runs"
    return root


def find_reusable_node(
    store: WorkflowRunStore,
    *,
    cache_identity: str,
    catalog: Any = None,
    verify_integrity: bool = True,
) -> NodeRun | None:
    """Search prior runs for a node with the identical cache identity whose
    outputs are still catalog-resolvable (and optionally integrity-verified).

    V8 M7: candidates come from the store's cache index (newest first) —
    only those runs are deserialized, instead of the previous full-history
    JSON rescan per lookup. Every candidate still passes the full reuse
    contract below. Returns the *prior* NodeRun (its receipt and output
    version ids are the reusable facts) or None. Corrupted/missing/deleted
    outputs never reuse.
    """
    for run_id, node_id in store.candidates_for_identity(cache_identity):
        try:
            run = store.load(run_id)
        except Exception:  # corrupted candidate — skip, keep looking
            logger.warning("cache candidate run %s unreadable; skipping", run_id)
            continue
        if run.state not in (RunState.COMPLETED, RunState.FAILED, RunState.INTERRUPTED):
            continue  # a still-running execution is not reusable evidence
        node_run = run.node_runs.get(node_id)
        if node_run is None:
            continue
        if node_run.state.value != "succeeded" or not node_run.cache_identity:
            continue
        if node_run.cache_identity != cache_identity:
            continue
        if node_run.from_cache:
            # Reuse chains to the *first* real execution; walking further
            # would compound stale-output risk for no benefit.
            continue
        if not node_run.output_version_ids:
            continue
        if catalog is not None and not _outputs_resolvable(
            catalog, node_run.output_version_ids, verify_integrity=verify_integrity
        ):
            continue
        return node_run
    return None


def run_lineage(store: WorkflowRunStore, run_id: str) -> list[str]:
    """Ancestral chain of runs (rerun/resume derivation), oldest first.

    V8 M7: [original, ..., run_id]. A malformed (cyclic) chain stops,
    never loops.
    """
    chain: list[str] = []
    seen: set[str] = set()
    current: str | None = str(run_id)
    while current is not None and current not in seen:
        seen.add(current)
        chain.append(current)
        try:
            run = store.load(current)
        except Exception:
            break
        current = run.parent_run_id
    chain.reverse()
    return chain


def _outputs_resolvable(catalog: Any, version_ids: tuple[str, ...], *, verify_integrity: bool) -> bool:
    for version_id in version_ids:
        try:
            ref = catalog.resolve_version(version_id)
        except Exception:
            return False
        if ref is None:
            return False
        if getattr(ref, "trashed", False):
            return False
        if verify_integrity and getattr(catalog, "verify_integrity", None) is not None:
            try:
                status = catalog.verify_integrity(version_id)
            except Exception:
                return False
            if getattr(status, "name", str(status)).upper() not in ("VERIFIED",):
                return False
    return True

File: libs/workflow_engine/src/store.cpp
// CONV-32: store.cpp — faithful port of paleo_workbench/workflow/dag/store.py.
// Atomic JSON checkpoints (store_version envelope, indent=1, no trailing
// newline), the in-memory cache index (rebuilt oldest-to-newest, read
// newest-first, maintained incrementally on save), find_reusable_node and
// the lineage walk. Semantics frozen against
// tools/oracle/generate_workflow_store_fixtures.py.
#include <pwb/workflow_engine/store.hpp>

#include <algorithm>
#include <atomic>
#include <chrono>
#include <fstream>
#include <sstream>
#include <system_error>
#include <utility>

#if defined(_WIN32)
#include <process.h>
#else
#include <unistd.h>
#endif

namespace pwb::workflow_engine {

namespace {

double wall_clock_seconds() {
    return std::chrono::duration<double>(
               std::chrono::system_clock::now().time_since_epoch())
        .count();
}

long long current_pid() {
#if defined(_WIN32)
    return _getpid();
#else
    return ::getpid();
#endif
}

unsigned long next_tmp_counter() {
    static std::atomic<unsigned long> counter{0};
    return counter.fetch_add(1);
}

// Python repr() of a short run id: single quotes, switching to double
// quotes only when the text holds ' but no " (mirrors CPython repr).
std::string python_repr_string(const std::string& text) {
    const bool has_single = text.find('\'') != std::string::npos;
    const bool has_double = text.find('"') != std::string::npos;
    const char quote = (has_single && !has_double) ? '"' : '\'';
    std::string out(1, quote);
    for (const char ch : text) {
        if (ch == '\\') {
            out += "\\\\";
        } else if (ch == '\n') {
            out += "\\n";
        } else if (ch == '\r') {
            out += "\\r";
        } else if (ch == '\t') {
            out += "\\t";
        } else if (ch == quote) {
            out += '\\';
            out += ch;
        } else {
            out += ch;
        }
    }
    out += quote;
    return out;
}

// getattr(run, "updated_at", None) or 0.0 — null/absent reads as 0.0
// (Python's `or` also maps a falsy 0.0 to 0.0, same value either way).
double updated_or_zero(const workflow_spec::WorkflowRun& run) {
    return run.updated_at.is_number() ? run.updated_at.get<double>() : 0.0;
}

// Sink bookkeeping for the FREE functions (find_reusable_node /
// run_lineage): store.hpp keeps sink_ private with no friend declaration,
// so set_log_sink mirrors the sink into an address-keyed registry the free
// functions can reach. The constructor erases any stale entry left by a
// destroyed store that happened to land on the same address, so a fresh
// store never inherits a dangling sink.
struct SinkRegistry {
    std::mutex mutex;
    std::unordered_map<const WorkflowRunStore*,
                       std::function<void(const std::string&)>>
        sinks;
};

SinkRegistry& sink_registry() {
    static SinkRegistry registry;
    return registry;
}

void log_for_store(const WorkflowRunStore* store, const std::string& message) {
    std::function<void(const std::string&)> sink;
    {
        SinkRegistry& registry = sink_registry();
        std::lock_guard<std::mutex> lock(registry.mutex);
        const auto it = registry.sinks.find(store);
        if (it == registry.sinks.end()) return;
        sink = it->second;
    }
    if (sink) sink(message);
}

void remember_sink(const WorkflowRunStore* store,
                   std::function<void(const std::string&)> sink) {
    SinkRegistry& registry = sink_registry();
    std::lock_guard<std::mutex> lock(registry.mutex);
    if (sink) {
        registry.sinks[store] = std::move(sink);
    } else {
        registry.sinks.erase(store);
    }
}

void forget_sink(const WorkflowRunStore* store) {
    SinkRegistry& registry = sink_registry();
    std::lock_guard<std::mutex> lock(registry.mutex);
    registry.sinks.erase(store);
}

std::string ascii_upper(std::string text) {
    for (char& ch : text) {
        if (ch >= 'a' && ch <= 'z') ch = static_cast<char>(ch - 'a' + 'A');
    }
    return text;
}

// store._outputs_resolvable: every recorded output version must resolve,
// not be trashed, and (when the catalog verifies integrity) be VERIFIED.
// NOTE: store.hpp declares the CatalogLike seam methods NON-const (Python
// duck-typing has no constness) while find_reusable_node receives a const
// pointer — the lookup path is logically non-mutating, so the seam is
// invoked through a const_cast.
bool outputs_resolvable(CatalogLike& catalog,
                        const std::vector<std::string>& version_ids,
                        bool verify_integrity) {
    for (const std::string& id : version_ids) {
        std::optional<VersionRefLike> ref;
        try {
            ref = catalog.resolve_version(id);
        } catch (...) {
            return false;
        }
        if (!ref.has_value()) return false;
        if (ref->trashed) return false;
        if (verify_integrity) {
            // nullopt == the catalog has no verify_integrity attribute.
            std::optional<std::string> status;
            try {
                status = catalog.verify_integrity(id);
            } catch (...) {
                return false;
            }
            if (status.has_value() && ascii_upper(*status) != "VERIFIED") {
                return false;
            }
        }
    }
    return true;
}

}  // namespace

// --------------------------------------------------------- error types --

RunNotFound::RunNotFound(const std::string& run_id,
                         const std::filesystem::path& root)
    : StoreError("no workflow run " + python_repr_string(run_id) + " in " +
                 root.string()) {}

CorruptCheckpoint::CorruptCheckpoint(const std::string& run_id,
                                     const std::string& parser_detail)
    : StoreError("workflow run " + python_repr_string(run_id) +
                 " checkpoint is corrupted: " + parser_detail) {}

// --------------------------------------------------------------- store --

WorkflowRunStore::WorkflowRunStore(std::filesystem::path root, Clock clock)
    : root_(std::move(root)), clock_(std::move(clock)) {
    std::error_code ec;
    std::filesystem::create_directories(root_, ec);  // mkdir -p
    forget_sink(this);
}

const std::filesystem::path& WorkflowRunStore::root() const noexcept {
    return root_;
}

std::filesystem::path WorkflowRunStore::save(workflow_spec::WorkflowRun& run) {
    run.updated_at = domain::Json(clock_ ? clock_() : wall_clock_seconds());
    // store._path: NO sanitization of run_id (parity-first; run ids are
    // engine-generated hex, so traversal is out of scope by design).
    const std::filesystem::path path = root_ / ("run-" + run.run_id + ".json");
    const domain::Json document = run.to_dict();  // stamped after update
    domain::Json payload = domain::Json::object();
    payload["store_version"] = 1;  // {"store_version": 1, **run.to_dict()}
    for (const auto& [key, value] : document.items()) {
        payload[key] = value;
    }
    // tempfile.mkstemp(dir=root, prefix=".tmp-run-", suffix=".json")
    // stand-in: pid + process-local counter keeps concurrent writers apart.
    const std::filesystem::path tmp =
        root_ / (".tmp-run-" + std::to_string(current_pid()) + "-" +
                 std::to_string(next_tmp_counter()) + ".json");
    try {
        {
            std::ofstream out(tmp, std::ios::binary | std::ios::trunc);
            if (!out) {
                throw StoreError("cannot open " + tmp.string() + " for writing");
            }
            // json.dump(..., ensure_ascii=False, indent=1) — no newline.
            out << payload.dump(1, ' ', false);
            out.flush();
            if (!out) {
                throw StoreError("cannot write " + tmp.string());
            }
        }
        // os.replace: POSIX rename over an existing checkpoint is atomic.
        std::filesystem::rename(tmp, path);
    } catch (...) {
        std::error_code ec;
        std::filesystem::remove(tmp, ec);  // best-effort unlink, then rethrow
        throw;
    }
    cache_index_entry(run);
    return path;
}

workflow_spec::WorkflowRun WorkflowRunStore::load(const std::string& run_id) const {
    const std::filesystem::path path = root_ / ("run-" + run_id + ".json");
    std::error_code ec;
    if (!std::filesystem::exists(path, ec)) {
        throw RunNotFound(run_id, root_);
    }
    std::ostringstream buffer;
    {
        std::ifstream in(path, std::ios::binary);
        buffer << in.rdbuf();
    }
    domain::Json data;
    try {
        data = domain::Json::parse(buffer.str());
    } catch (const std::exception& exc) {
        // A corrupted checkpoint is never executed from.
        throw CorruptCheckpoint(run_id, exc.what());
    }
    return workflow_spec::WorkflowRun::from_dict(data);
}

std::vector<std::string> WorkflowRunStore::list_run_ids() const {
    static constexpr std::string_view kPrefix = "run-";
    static constexpr std::string_view kSuffix = ".json";
    std::vector<std::string> ids;
    std::error_code ec;
    // root.glob("run-*.json"): dot-prefixed tmp files never match.
    for (const auto& entry : std::filesystem::directory_iterator(root_, ec)) {
        const std::string name = entry.path().filename().string();
        if (name.size() < kPrefix.size() + kSuffix.size()) continue;
        if (!name.starts_with(kPrefix) || !name.ends_with(kSuffix)) continue;
        ids.push_back(name.substr(kPrefix.size(),
                                  name.size() - kPrefix.size() - kSuffix.size()));
    }
    std::sort(ids.begin(), ids.end());
    return ids;
}

std::vector<workflow_spec::WorkflowRun> WorkflowRunStore::list_runs() const {
    std::vector<workflow_spec::WorkflowRun> runs;
    for (const std::string& run_id : list_run_ids()) {
        try {
            runs.push_back(load(run_id));
        } catch (...) {  // a corrupted run must not poison cache lookups
            if (sink_) sink_("skipping unreadable workflow run " + run_id);
        }
    }
    return runs;
}

void WorkflowRunStore::cache_index_entry(
    const workflow_spec::WorkflowRun& run) {
    if (!index_built_) return;  // Python: _cache_index is None -> no-op
    if (run.state == workflow_spec::RunState::running) {
        return;  // not reusable evidence until terminal
    }
    std::lock_guard<std::mutex> lock(index_mutex_);
    for (const auto& [node_id, node_run] : run.node_runs) {
        if (to_string(node_run.state) != "succeeded") continue;
        // `node_run.cache_identity` truthy: present AND non-empty.
        if (!node_run.cache_identity.has_value() ||
            node_run.cache_identity->empty()) {
            continue;
        }
        if (node_run.from_cache) continue;  // reuse chains to first execution
        if (node_run.output_version_ids.empty()) continue;
        const std::string pair_key = run.run_id + '\x1f' + node_id;
        if (indexed_pairs_.count(pair_key) != 0) continue;
        indexed_pairs_.insert(pair_key);
        cache_index_[*node_run.cache_identity].emplace_back(run.run_id, node_id);
    }
}

int WorkflowRunStore::rebuild_cache_index() {
    {
        std::lock_guard<std::mutex> lock(index_mutex_);
        cache_index_.clear();
        indexed_pairs_.clear();
        index_built_ = true;
    }
    std::vector<workflow_spec::WorkflowRun> runs = list_runs();
    // Oldest to newest by (updated_at or 0.0), stable — the newest-first
    // read is real recency, not a lexicographic accident.
    std::stable_sort(runs.begin(), runs.end(),
                     [](const workflow_spec::WorkflowRun& a,
                        const workflow_spec::WorkflowRun& b) {
                         return updated_or_zero(a) < updated_or_zero(b);
                     });
    for (const auto& run : runs) {
        cache_index_entry(run);
    }
    std::lock_guard<std::mutex> lock(index_mutex_);
    int total = 0;
    for (const auto& [identity, entries] : cache_index_) {
        total += static_cast<int>(entries.size());
    }
    return total;
}

std::vector<std::pair<std::string, std::string>>
WorkflowRunStore::candidates_for_identity(const std::string& cache_identity) {
    if (!index_built_) rebuild_cache_index();  // lazy build
    std::vector<std::pair<std::string, std::string>> entries;
    {
        std::lock_guard<std::mutex> lock(index_mutex_);
        const auto it = cache_index_.find(cache_identity);
        if (it != cache_index_.end()) entries = it->second;
    }
    std::reverse(entries.begin(), entries.end());  // newest first
    return entries;
}

void WorkflowRunStore::set_log_sink(
    std::function<void(const std::string&)> sink) {
    sink_ = sink;
    remember_sink(this, std::move(sink));
}

// -------------------------------------------------------- default root --

std::filesystem::path default_store_root(const std::string& project_path) {
    if (project_path.empty()) {
        return default_store_root();
    }
    // paths.artifact_dir_for: strip ".paleo.json" from the FILENAME and
    // append ".artifacts" as a sibling directory.
    const std::filesystem::path project(project_path);
    std::string name = project.filename().string();
    static constexpr std::string_view kSuffix = ".paleo.json";
    if (name.size() >= kSuffix.size() &&
        name.compare(name.size() - kSuffix.size(), kSuffix.size(), kSuffix) ==
            0) {
        name.resize(name.size() - kSuffix.size());
    }
    return (project.parent_path() / (name + ".artifacts")) / "workflows";
}

std::filesystem::path default_store_root() {
    return std::filesystem::temp_directory_path() / "paleo-workflow-runs";
}

// ----------------------------------------------------- reuse + lineage --

std::optional<workflow_spec::NodeRun> find_reusable_node(
    const WorkflowRunStore& store, const std::string& cache_identity,
    const CatalogLike* catalog, bool verify_integrity) {
    // candidates_for_identity lazily builds the index (mutating cache
    // state) but is observationally a lookup; store.hpp models the Python
    // seam without const, hence the cast.
    WorkflowRunStore& mutable_store = const_cast<WorkflowRunStore&>(store);
    for (const auto& [run_id, node_id] :
         mutable_store.candidates_for_identity(cache_identity)) {
        workflow_spec::WorkflowRun run;
        try {
            run = store.load(run_id);
        } catch (...) {
            log_for_store(&store,
                          "cache candidate run " + run_id + " unreadable; skipping");
            continue;
        }
        using workflow_spec::RunState;
        if (run.state != RunState::completed && run.state != RunState::failed &&
            run.state != RunState::interrupted) {
            continue;  // a still-running execution is not reusable evidence
        }
        const workflow_spec::NodeRun* node_run = run.find_node_run(node_id);
        if (node_run == nullptr) continue;
        if (to_string(node_run->state) != "succeeded" ||
            !node_run->cache_identity.has_value() ||
            node_run->cache_identity->empty()) {
            continue;
        }
        if (*node_run->cache_identity != cache_identity) continue;
        if (node_run->from_cache) continue;
        if (node_run->output_version_ids.empty()) continue;
        if (catalog != nullptr &&
            !outputs_resolvable(*const_cast<CatalogLike*>(catalog),
                                node_run->output_version_ids,
                                verify_integrity)) {
            continue;
        }
        return *node_run;
    }
    return std::nullopt;
}

std::vector<std::string> run_lineage(const WorkflowRunStore& store,
                                     const std::string& run_id) {
    std::vector<std::string> chain;
    std::unordered_set<std::string> seen;
    std::optional<std::string> current = run_id;
    while (current.has_value() && seen.count(*current) == 0) {
        seen.insert(*current);
        chain.push_back(*current);
        try {
            const workflow_spec::WorkflowRun run = store.load(*current);
            current = run.parent_run_id;
        } catch (...) {
            break;  // unreadable link stops the walk, keeping what walked
        }
    }
    std::reverse(chain.begin(), chain.end());  // oldest first
    return chain;
}

}  // namespace pwb::workflow_engine
